README
NAME
    PITA::Scheme - PITA Testing Schemes

SYNOPSIS
      # Have the scheme load up from the provided config
      my $scheme = PITA::Scheme->new(
            injector => $injector,
            workarea => $workarea,
            );
  
      # Prepare to run the tests
      $scheme->prepare_all;
  
      # Run the tests
      $scheme->execute_all;

DESCRIPTION
    While most of the PITA system exists outside the guest testing images
    and tries to have as little interaction with them as possible, there is
    one part that needs to be run from inside it.

    PITA::Scheme objects live inside the image and does three main tasks.

    1. Unpack the package and prepare the testing environment

    2. Run the sequence of commands to execute the tests and capture the
    results.

    3. Package the results as a PITA::XML::Report and send it to the
    PITA::Host::ResultServer.

    This functionality is implemented in a module structure that is highly
    subclassable. In this way, PITA can support multiple different testing
    schemes for multiple different languages and installer types.

Setting up a Testing Image
    Each image that will be set up will require a bit of customisation, as
    the entire point of this type of testing is that every environment is
    different.

    However, by keeping most of the functionality in the PITA::Scheme
    objects, all you should need to do is to arrange for a simple Perl
    script to be launched, that feeds some initial configuration to the
    PITA::Scheme object.

    And it should do the rest. Or die... but we'll cover that later.

METHODS
    Please excuse the lack of details for now...

    TO BE COMPLETED

SUPPORT
    Bugs should be reported via the CPAN bug tracker at

    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=PITA-Scheme>

    For other issues, contact the author.

AUTHOR
    Adam Kennedy <cpan@ali.as>, <http://ali.as/>

SEE ALSO
    The Perl Image Testing Architecture (<http://ali.as/pita/>)

    PITA, PITA::XML, PITA::Host::ResultServer

COPYRIGHT
    Copyright 2005 Adam Kennedy. All rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.



Changes
Revision history for Perl extension PITA::Scheme

0.20  Thu  2 Feb 2006
	- Fixed a minor cleanup issue, we need to undef the pushd earlier
	- Synchronised version with PITA 0.20 release

0.15  Tue 31 Jan 2006
	- Completing refactoring for the upcoming PITA 0.20 release

0.14  Mon 30 Jan 2006
	- Adding support for detecting Perl5 platforms

0.13  Mon 23 Jan 2006
	- Make sure we use File::pushd so we don't end up stuck down
	  some temp directory.

0.12  Sun 22 Jan 2006
	- Upgrading to properly support the Request ids in PITA-XML 0.15

0.11  Sun 15 Jan 2006
	- Added a param validation method _DRIVER($scheme)
	- Added the AUTOMATED_TESTING and PERL_AUTOINSTALL flags

0.10  Sat 14 Jan 2006
	- Converted over to the renamed PITA::XML

0.09  Mon  9 Jan 2006
	- Ripped out all the HTTP transfer stuff, as it has been moved
	  over to PITA-Image.
	- Take less configuration, only the scheme-specific file.
	- Requires more constructor arguments.

0.08  Sat  7 Jan 2006
	- Converted command execution code from flakey IPC::Cmd to the
	  much saner IPC::Run3.

0.07  ###############
	- Fixed misspelled method name in Scheme.pm

0.06  Tue  3 Jan 2006
	- Added the concept of request identifiers, because embedding
	  that information inside the ResultServer isn't going to work
	  once we add more functionality and make it a SupportServer.

0.05  Tue  3 Jan 2006
	- One of the tests need Mobule::Build to be available

0.04  Tue  3 Jan 2006
	- Accidentally left a Build file in Brian the test dummy.
	  This was causing test problems

0.03  Mon  2 Jan 2006
	- Module::Build manual suggests more compatible in more places to
	  use C<perl Build>/etc than C<./Build>
	- Incrementing PITA::XML dep

0.02  Mon  2 Jan 2006
	- Completed what should be the first usable version
	- Completed Scheme modules for perl5, perl5.make, perl5.build
	- Should be able to put the results to the server if told where.

0.01  Sat 31 Dec 2005
	- Initial implementation
	- Completes sequence up to prepare_package
	- Releasing to get some early CPAN Testers results
	  (As this needs to be highly platform neutral)


lib/PITA/Scheme.pm
package PITA::Scheme;

=pod

=head1 NAME

PITA::Scheme - PITA Testing Schemes

=head1 SYNOPSIS

  # Have the scheme load up from the provided config
  my $scheme = PITA::Scheme->new(
	injector => $injector,
	workarea => $workarea,
	);
  
  # Prepare to run the tests
  $scheme->prepare_all;
  
  # Run the tests
  $scheme->execute_all;

=head1 DESCRIPTION

While most of the PITA system exists outside the guest testing images and
tries to have as little interaction with them as possible, there is one
part that needs to be run from inside it.

PITA::Scheme objects live inside the image and does three main tasks.

1. Unpack the package and prepare the testing environment

2. Run the sequence of commands to execute the tests and capture
the results.

3. Package the results as a L<PITA::XML::Report> and send it to the
L<PITA::Host::ResultServer>.

This functionality is implemented in a module structure that is highly
subclassable. In this way, L<PITA> can support multiple different
testing schemes for multiple different languages and installer types.

=head1 Setting up a Testing Image

Each image that will be set up will require a bit of customisation,
as the entire point of this type of testing is that every environment
is different.

However, by keeping most of the functionality in the L<PITA::Scheme>
objects, all you should need to do is to arrange for a simple Perl
script to be launched, that feeds some initial configuration to the
L<PITA::Scheme> object.

And it should do the rest. Or die... but we'll cover that later.

=head1 METHODS

Please excuse the lack of details for now...

TO BE COMPLETED

=cut

use 5.005;
use strict;
use Carp         ();
use IPC::Run3    ();
use File::Spec   ();
use Params::Util '_INSTANCE',
                 '_POSINT',
                 '_STRING',
                 '_ARRAY',
                 '_CLASS';
use PITA::XML    ();

use vars qw{$VERSION};
BEGIN {
	$VERSION = '0.20';
}





#####################################################################
# Constructor

sub new {
	my $class = shift;
	my $self  = bless { @_ }, $class;

	# Apply the default path if needed
	unless ( $self->path ) {
		$self->{path} = $self->default_path;
	}

	# Cursory checking for compulsory params
	foreach my $param ( qw{injector workarea scheme path} ) {
		next if $self->$param();
		Carp::croak("Missing compulsory param '$param'");
	}

	# Load the request from a file if needed
	unless ( $self->request ) {
		$self->{request_xml} = File::Spec->catfile( $self->injector, $self->request_xml );
		unless ( -f $self->request_xml and -r _ ) {
			Carp::croak('Missing request file, or no permissions');
		}
		$self->{request} = PITA::XML::Request->read( $self->request_xml );
	}
	unless ( _INSTANCE($self->request, 'PITA::XML::Request') ) {
		Carp::croak("Bad report Request or failed to load one from "
			. $self->request_xml );
	}
	unless ( $self->request->scheme eq $self->scheme ) {
		Carp::croak("Test scheme in image.conf does not match Request scheme");
	}

	# Check the request identifier
	unless ( _POSINT($self->request_id) ) {
		Carp::croak("Missing or bad request_id for this test instance");
	}

	$self;
}





#####################################################################
# Accessors and convience methods

sub injector {
	$_[0]->{injector};
}

sub workarea {
	$_[0]->{workarea};
}

sub scheme {
	$_[0]->{scheme};
}

sub path {
	$_[0]->{path};
}

sub request_xml {
	$_[0]->{request_xml};
}

sub request {
	$_[0]->{request};
}

sub request_id {
	my $self = shift;
	if ( $self->request and $self->request->id ) {
		# New style request with an id
		return $self->request->id;
	} else {
		# Manually passed request_id
		return $self->{request_id};
	}

	undef;
}

sub install {
	$_[0]->{install};	
}

sub report {
	$_[0]->{report};
}





#####################################################################
# PITA::Scheme Methods

sub load_config {
	my $self = shift;

	# Load the config file
	$self->{config} = Config::Tiny->new( $self->{config_file} )
		or Carp::croak("Failed to load config file: "
			. Config::Tiny->errstr);

	# Validate some basics

	1;
}

# Do the various preparations
sub prepare_all {
	my $self = shift;
	return 1 if $self->install;

	# Prepare the package
	$self->prepare_package;

	# Prepare the environment
	$self->prepare_environment;

	# Prepare the report
	$self->prepare_report;

	1;
}

# Nothing, yet
sub prepare_package {
	my $self = shift;
	1;
}

sub prepare_report {
	my $self = shift;
	return 1 if $self->install;

	# Create the install object
	$self->{install} = PITA::XML::Install->new(
		request  => $self->request,
		platform => $self->platform,
		);

	# Create the main report object
	$self->{report} ||= PITA::XML::Report->new;
	$self->report->add_install( $self->install );

	1;
}

sub execute_command {
	my $self = shift;
	my $cmd  = _ARRAY( [ @_ ] ) or Carp::croak(
		"execute_command not passed an ARRAY ref as command"
		);

	# Execute the command
	my $stdout  = '';
	my $stderr  = '';
	my $success = IPC::Run3::run3( $cmd, \undef, \$stdout, \$stderr );

	# Turn the results into a Command object
	my $command = PITA::XML::Command->new(
		cmd    => join( ' ', @$cmd ),
		stdout => \$stdout,
		stderr => \$stderr,
		);
	unless ( _INSTANCE($command, 'PITA::XML::Command') ) {
		Carp::croak("Error creating ::Command");
	}

	# If we have a PITA::XML::Install object available,
	# automatically add to it.
	if ( $self->install ) {
		$self->install->add_command( $command );
	}

	$command;
}





#####################################################################
# Support Methods

# Validate a usable scheme, returning a (loaded) driver class
sub _DRIVER {
	my $class  = shift;

	# Resolve the specific testing scheme class for this run
	my $scheme = shift;
	unless ( _STRING($scheme) ) {
		Carp::croak("Missing or invalid scheme");
	}
	my $driver = join( '::', 'PITA', 'Scheme', map { ucfirst $_ } split /\./, lc($scheme || '') );
	unless ( _CLASS($driver) ) {
		Carp::croak("Invalid scheme '$scheme'");
	}

	# Load the scheme class
	eval "require $driver;";
	if ( $@ =~ /^Can\'t locate PITA/ ) {
		Carp::croak("Scheme '$scheme' is unsupported on this Guest");
	} elsif ( $@ ) {
		Carp::croak("Error loading scheme '$scheme' driver $driver: $@");
	}

	$driver;
}

1;

=pod

=head1 SUPPORT

Bugs should be reported via the CPAN bug tracker at

L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=PITA-Scheme>

For other issues, contact the author.

=head1 AUTHOR

Adam Kennedy E<lt>cpan@ali.asE<gt>, L<http://ali.as/>

=head1 SEE ALSO

The Perl Image Testing Architecture (L<http://ali.as/pita/>)

L<PITA>, L<PITA::XML>, L<PITA::Host::ResultServer>

=head1 COPYRIGHT

Copyright 2005 Adam Kennedy. All rights reserved.

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut


inc/Module/Install.pm
#line 1 "/tmp/pita/PITA-Scheme/inc/Module/Install.pm - /usr/local/share/perl/5.8.4/Module/Install.pm"
package Module::Install;

use 5.004;
use strict 'vars';
use vars qw{$VERSION};
BEGIN {
    # Don't forget to update Module::Install::Admin too!
    $VERSION = '0.55';
}

# inc::Module::Install must be loaded first
unless ( $INC{join('/', inc => split(/::/, __PACKAGE__)).'.pm'} ) {
    die <<"END_DIE";
Please invoke ${\__PACKAGE__} with:

    use inc::${\__PACKAGE__};

not:

    use ${\__PACKAGE__};

END_DIE
}

use Cwd        ();
use FindBin;
use File::Find ();
use File::Path ();

*inc::Module::Install::VERSION = *VERSION;
@inc::Module::Install::ISA     = 'Module::Install';

sub autoload {
    my $self   = shift;
    my $caller = $self->_caller;
    my $cwd    = Cwd::cwd();
    my $sym    = "$caller\::AUTOLOAD";

    $sym->{$cwd} = sub {
        my $pwd = Cwd::cwd();
        if ( my $code = $sym->{$pwd} ) {
            # delegate back to parent dirs
            goto &$code unless $cwd eq $pwd;
        }
        $$sym =~ /([^:]+)$/ or die "Cannot autoload $caller - $sym";
        unshift @_, ($self, $1);
        goto &{$self->can('call')} unless uc($1) eq $1;
    };
}

sub import {
    my $class = shift;
    my $self  = $class->new(@_);

    unless ( -f $self->{file} ) {
        require "$self->{path}/$self->{dispatch}.pm";
        File::Path::mkpath("$self->{prefix}/$self->{author}");
        $self->{admin} = 
          "$self->{name}::$self->{dispatch}"->new(_top => $self);
        $self->{admin}->init;
        @_ = ($class, _self => $self);
        goto &{"$self->{name}::import"};
    }

    *{$self->_caller . "::AUTOLOAD"} = $self->autoload;
    $self->preload;

    # Unregister loader and worker packages so subdirs can use them again
    delete $INC{"$self->{file}"};
    delete $INC{"$self->{path}.pm"};
}

sub preload {
    my ($self) = @_;

        unless ( $self->{extentions} ) {
                $self->load_extensions(
                        "$self->{prefix}/$self->{path}", $self
                        );
        }

    my @exts = @{$self->{extensions}};
    unless ( @exts ) {
        my $admin = $self->{admin};
        @exts = $admin->load_all_extensions;
    }

    my %seen_method;
    foreach my $obj ( @exts ) {
        while (my ($method, $glob) = each %{ref($obj) . '::'}) {
            next unless defined *{$glob}{CODE};
            next if $method =~ /^_/;
            next if $method eq uc($method);
            $seen_method{$method}++;
        }
    }

    my $caller = $self->_caller;
    foreach my $name (sort keys %seen_method) {
        *{"${caller}::$name"} = sub {
            ${"${caller}::AUTOLOAD"} = "${caller}::$name";
            goto &{"${caller}::AUTOLOAD"};
        };
    }
}

sub new {
    my ($class, %args) = @_;

    # ignore the prefix on extension modules built from top level.
    my $base_path = Cwd::abs_path($FindBin::Bin);
    unless ( Cwd::abs_path(Cwd::cwd()) eq $base_path ) {
        delete $args{prefix};
    }

    return $args{_self} if $args{_self};

    $args{dispatch} ||= 'Admin';
    $args{prefix}   ||= 'inc';
    $args{author}   ||= '.author';
    $args{bundle}   ||= 'inc/BUNDLES';
    $args{base}     ||= $base_path;

    $class =~ s/^\Q$args{prefix}\E:://;
    $args{name}     ||= $class;
    $args{version}  ||= $class->VERSION;

    unless ($args{path}) {
        $args{path}  = $args{name};
        $args{path}  =~ s!::!/!g;
    }
    $args{file}     ||= "$args{base}/$args{prefix}/$args{path}.pm";

    bless(\%args, $class);
}

sub call {
    my $self   = shift;
    my $method = shift;
    my $obj    = $self->load($method) or return;

    unshift @_, $obj;
    goto &{$obj->can($method)};
}

sub load {
    my ($self, $method) = @_;

    $self->load_extensions(
        "$self->{prefix}/$self->{path}", $self
    ) unless $self->{extensions};

    foreach my $obj (@{$self->{extensions}}) {
        return $obj if $obj->can($method);
    }

    my $admin = $self->{admin} or die <<"END_DIE";
The '$method' method does not exist in the '$self->{prefix}' path!
Please remove the '$self->{prefix}' directory and run $0 again to load it.
END_DIE

    my $obj = $admin->load($method, 1);
    push @{$self->{extensions}}, $obj;

    $obj;
}

sub load_extensions {
    my ($self, $path, $top_obj) = @_;

    unshift @INC, $self->{prefix}
        unless grep { $_ eq $self->{prefix} } @INC;

    local @INC = ($path, @INC);
    foreach my $rv ($self->find_extensions($path)) {
        my ($file, $pkg) = @{$rv};
        next if $self->{pathnames}{$pkg};

        local $@;
        my $new = eval { require $file; $pkg->can('new') };
        unless ( $new ) {
            warn $@ if $@;
            next;
        }
        $self->{pathnames}{$pkg} = delete $INC{$file};
        push @{$self->{extensions}}, &{$new}($pkg, _top => $top_obj );
    }

    $self->{extensions} ||= [];
}

sub find_extensions {
    my ($self, $path) = @_;

    my @found;
    File::Find::find( sub {
        my $file = $File::Find::name;
        return unless $file =~ m!^\Q$path\E/(.+)\.pm\Z!is;
        return if $1 eq $self->{dispatch};

        $file = "$self->{path}/$1.pm";
        my $pkg = "$self->{name}::$1"; $pkg =~ s!/!::!g;
        push @found, [ $file, $pkg ];
    }, $path ) if -d $path;

    @found;
}

sub _caller {
    my $depth  = 0;
    my $caller = caller($depth);

    while ($caller eq __PACKAGE__) {
        $depth++;
        $caller = caller($depth);
    }

    $caller;
}

1;


lib/PITA/Scheme/Perl5.pm
package PITA::Scheme::Perl5;

# Auto-sensing Perl 5 scheme.
# Use Makefile.PL or Build.PL as appropriate.

# We won't know this until AFTER the extract_package method,
# so at the end of this method we will look for the appropriate
# file and rebless ourself into the class for it.

use 5.005;
use strict;
use base 'PITA::Scheme::Perl';
use File::Spec  ();
use File::Which ();

use vars qw{$VERSION};
BEGIN {
	$VERSION = '0.20';
}





#####################################################################
# Constructor

sub new {
	my $class = shift;
	my $self  = $class->SUPER::new(@_);
	$self;
}





#####################################################################
# PITA::Scheme Methods

sub default_path {
	File::Which::which('perl') || '';
}

sub prepare_package {
	my $self = shift;

	# Do the generic unpacking
	$self->SUPER::prepare_package(@_);

	# Use the Makefile.PL in preference to Build.PL as it should be
	# more mature than a Module::Build installer.
	# Whichever one we call, it will end up called the same
	# prepare_package in ::Perl. This isn't a problem, as it knows
	# how to shortcut.
	if ( -f $self->workarea_file('Makefile.PL') ) {
		require PITA::Scheme::Perl5::Make;
		bless $self, 'PITA::Scheme::Perl5::Make';
		return $self->prepare_package;

	} elsif ( -f $self->workarea_file('Build.PL') ) {
		require PITA::Scheme::Perl5::Make;
		bless $self, 'PITA::Scheme::Perl5::Make';
		return $self->prepare_package;

	}
	
	# Doesn't have either
	Carp::croak("Perl5 package contains neither Makefile.PL or Build.PL");
}

1;


lib/PITA/Scheme/Perl5/Make.pm
package PITA::Scheme::Perl5::Make;

# Class for implementing the perl5-make testing scheme

use 5.005;
use strict;
use base 'PITA::Scheme::Perl';
use Carp        ();
use File::Which ();

use vars qw{$VERSION};
BEGIN {
	$VERSION = '0.20';
}





#####################################################################
# Constructor

sub default_path {
	File::Which::which('perl') || '';
}

sub new {
	my $class = shift;
	my $self  = $class->SUPER::new(@_);

	### Additional checks

	$self;
}





#####################################################################
# PITA::Scheme Methods

sub prepare_package {
	my $self = shift;

	# Do the generic unpacking
	$self->SUPER::prepare_package(@_);

	# Validate that the package has a Makefile.PL in the root
	unless ( -f $self->workarea_file('Makefile.PL') ) {
		Carp::croak("Package does not contain a Makefile.PL");
	}

	$self;
}

sub execute_all {
	my $self = shift;

	# Run the Makefile.PL
	$self->execute_makefilepl or return '';

	# Run the make
	$self->execute_make or return '';

	# Run the tests
	$self->execute_maketest or return '';

	1;
}

sub execute_makefilepl {
	my $self = shift;
	unless ( -f $self->workarea_file('Makefile.PL') ) {
		Carp::croak("Cannot execute_makefilepl without a Makefile.PL");
	}

	# Run the Makefile.PL
	my $command = $self->execute_command('perl', 'Makefile.PL');

	# Did it create a make file
	if ( -f $self->workarea_file('Makefile') ) {
		# Worked as expected
		### Do we need to add stuff here later?
		return 1;
	}

	# Didn't work
	### Do we need to add stuff here later?
	return '';
}

sub execute_make {
	my $self = shift;
	unless ( -f $self->workarea_file('Makefile') ) {
		Carp::croak("Cannot execute_make without a Makefile");
	}

	# Run the make
	my $command = $self->execute_command('make');

	# Did it create a blib directory?
	if ( -d $self->workarea_file('blib') ) {
		# Worked as expected
		### Do we need to add stuff here later?
		return 1;
	}

	# Didn't work
	### Do we need to add stuff here later?
	return '';
}

sub execute_maketest {
	my $self = shift;
	unless ( -f $self->workarea_file('Makefile') ) {
		Carp::croak("Cannot execute_maketest without a Makefile");
	}
	unless ( -d $self->workarea_file('blib') ) {
		Carp::croak("Cannot execute_maketest without a blib");
	}

	# Run the make test
	my $command = $self->execute_command('make', 'test');

	# Did it... erm...
	if ( 1 ) {
		# Worked as expected
		### Do we need to add stuff here later?
		return 1;
	}

	# Didn't work
	### Do we need to add stuff here later?
	return '';
}

1;


lib/PITA/Scheme/Perl5/Build.pm
package PITA::Scheme::Perl5::Build;

# Class for implementing the perl5-build testing scheme

use 5.005;
use strict;
use base 'PITA::Scheme::Perl';
use Carp        ();
use File::Spec  ();
use File::Which ();

use vars qw{$VERSION};
BEGIN {
	$VERSION = '0.20';
}





#####################################################################
# Constructor

sub default_path {
	File::Which::which('perl') || '';
}

sub new {
	my $class = shift;
	my $self  = $class->SUPER::new(@_);

	### Additional checks, if any

	$self;
}





#####################################################################
# PITA::Scheme Methods

sub prepare_package {
	my $self = shift;

	# Do the generic unpacking
	$self->SUPER::prepare_package(@_);

	# Validate that the package has a Makefile.PL in the root
	unless ( -f $self->workarea_file('Build.PL') ) {
		Carp::croak("Package does not contain a Makefile.PL");
	}

	$self;
}

sub execute_all {
	my $self = shift;

	# Run the Makefile.PL
	$self->execute_buildpl or return '';

	# Run the make
	$self->execute_build or return '';

	# Run the tests
	$self->execute_buildtest or return '';

	1;
}

sub execute_buildpl {
	my $self = shift;
	unless ( -f $self->workarea_file('Build.PL') ) {
		Carp::croak("Cannot execute_makefilepl without a Build.PL");
	}

	# Run the Makefile.PL
	my $command = $self->execute_command('perl', 'Build.PL');

	# Did it create a make file
	if ( -f $self->workarea_file('Build') ) {
		# Worked as expected
		### Do we need to add stuff here later?
		return 1;
	}

	# Didn't work
	### Do we need to add stuff here later?
	return '';
}

sub execute_build {
	my $self = shift;
	unless ( -f $self->workarea_file('Build') ) {
		Carp::croak("Cannot execute_make without a Build file");
	}

	# Run the make
	my $command = $self->execute_command('perl', 'Build');

	# Did it create a blib directory?
	if ( -d $self->workarea_file('blib') ) {
		# Worked as expected
		### Do we need to add stuff here later?
		return 1;
	}

	# Didn't work
	### Do we need to add stuff here later?
	return '';
}

sub execute_buildtest {
	my $self = shift;
	unless ( -f $self->workarea_file('Build') ) {
		Carp::croak("Cannot execute_maketest without a Build file");
	}
	unless ( -d $self->workarea_file('blib') ) {
		Carp::croak("Cannot execute_maketest without a blib");
	}

	# Run the make test
	my $command = $self->execute_command('perl', 'Build', 'test');

	# Did it... erm...
	if ( 1 ) {
		# Worked as expected
		### Do we need to add stuff here later?
		return 1;
	}

	# Didn't work
	### Do we need to add stuff here later?
	return '';
}

1;


lib/PITA/Scheme/Perl/Discovery.pm
package PITA::Scheme::Perl::Discovery;

# Provides a mechanism for discovering the platform of an arbitrary
# perl interpreter, by using Process::Delegatable.

use 5.005;
use strict;
use base 'Process::Delegatable',
         'Process::Storable',
         'Process';
use Params::Util '_STRING';
use File::Spec   ();
use PITA::XML    ();

use vars qw{$VERSION};
BEGIN {
	$VERSION = '0.20';
}





#####################################################################
# Constructor and Accessors

sub new {
	my $class = shift;

	# Create the object
	my $self = bless { @_ }, $class;

	# Check that the program exists.
	# The path should not be the command to call, it should be
	# the actual file path of the Perl executable.
	unless ( _STRING($self->{path}) ) {
		Carp::croak("Did not provide a path to perl");
	}
	unless ( File::Spec->file_name_is_absolute($self->{path}) ) {
		Carp::croak("The path $self->{path} is not absolute");
	}
	unless ( -f $self->{path} ) {
		Carp::croak("The path $self->{path} does not exist");
	}
	unless ( -x $self->{path} ) {
		Carp::croak("The path $self->{path} is not executable");
	}

	# Check that it is indeed Perl
	my @output = `$self->{path} -v`;
	chomp @output;
	shift @output if $output[0] =~ /^\s*$/;
	unless ( $output[0] =~ /^This is perl,/ ) {
		Carp::croak("The path $self->{path} is not Perl");
	}

	$self;
}

sub path {
	$_[0]->{path};
}

sub platform {
	$_[0]->{platform};
}





#####################################################################
# Process::Delegatable

# Allow delegate to be called without a param.
# And shortcut if already run.
sub delegate {
	my $self = shift;
	return 1 if defined $self->platform;
	$self->SUPER::delegate($self->path, @_);
}





#####################################################################
# Process Methods

sub prepare {
	# We are already running inside the Perl interpreter we want
	# to capture at this point, and we have already loaded
	# PITA::XML::Platform, so there's nothing else we need to do.
	return 1;
}

sub run {
	my $self = shift;

	# Do the autodetection
	my $platform = eval { PITA::XML::Platform->autodetect_perl5 };
	if ( $platform ) {
		$self->{platform} = $platform;
	} else {
		$self->{errstr}   = $@;
	}

	1;
}





#####################################################################
# Error Methods

sub errstr {
	$_[0]->{errstr};
}

1;


lib/PITA/Scheme/Perl.pm
package PITA::Scheme::Perl;

# Base class for all schemes working with Perl-like distributions.
# Provides bits of common functionality.

use 5.005;
use strict;
use base 'PITA::Scheme';
use Carp             ();
use File::Spec       ();
use File::pushd      ();
use Params::Util     '_INSTANCE';
use Archive::Extract ();

use vars qw{$VERSION};
BEGIN {
	$VERSION = '0.20';
}





#####################################################################
# Generic Constructor

# Do the extra common checks we couldn't do in the main class
sub new {
	my $class = shift;
	my $self  = $class->SUPER::new(@_);

	# Can we locate the package?
	my $filename = $self->request->filename;
	$self->{archive} = File::Spec->catfile( $self->injector, $filename );
	unless ( -f $self->{archive} ) {
		Carp::croak('Failed to find package $filename in injector');
	}

	$self;
}

sub archive {
	$_[0]->{archive};
}

sub extract_path {
	$_[0]->{extract_path};
}

sub extract_files {
	my $files = $_[0]->{extract_files};
	$files ? @$files : ();
}

sub platform {
	$_[0]->{platform};
}





#####################################################################
# PITA::Scheme Methods

sub prepare_package {
	my $self = shift;
	$self->SUPER::prepare_package(@_);
	return 1 if $self->{extract_files};

	# Extract the package to the working directory
	my $archive = Archive::Extract->new( archive => $self->archive )
		or Carp::croak("Package is not an archive, or not extractable");

	# Extract the archive to the working directory
	local $Archive::Extract::WARN = 0;
	my $ok = $archive->extract( to => $self->workarea )
		or Carp::croak("Error extracting package: "
			. $archive->error );

	# Save the list of files
	$self->{extract_path}  = $archive->extract_path;
	$self->{extract_files} = $archive->files;
	### For now this list is unreliable and inconsistent.

	# Look for a single subdirectory and descend if needed
	$self;
}

sub prepare_environment {
	my $self = shift;
	unless ( $self->extract_path ) {
		Carp::croak("Cannot call prepare_environment without extracting pachage");
	}

	# Change to the extraction directory
	unless ( $self->{_chdir} ) {
		unless ( $self->{_chdir} = File::pushd::pushd( $self->extract_path ) ) {
			Carp::croak("Failed to change to extract_path for execution");
		}
	}

	# Set any general environment variables
	$ENV{PERL_AUTOINSTALL}  = '--defaultdeps';
	$ENV{AUTOMATED_TESTING} = ref($self);

	# Set any environment variables from the Request
	### Not supported by PITA::XML::Request yet

	# Save the platform configuration
	$self->{platform} = PITA::XML::Platform->autodetect_perl5;
	unless ( _INSTANCE($self->{platform}, 'PITA::XML::Platform') ) {
		Carp::croak("Failed to capture platform information");
	}

	1;
}





#####################################################################
# PITA::Scheme::Perl Methods

# Mainly a convenience for now.
sub workarea_file {
	my $self = shift;

	# If the package has been extracted, prefer its
	# interpretation of being where the workarea is.
	my $workarea = defined $self->extract_path
		? $self->extract_path
		: $self->workarea;
	File::Spec->catfile( $workarea, shift );
}





#####################################################################
# Support Methods

sub DESTROY {
	# Remove the _chdir explicitly early
	if ( defined $_[0]->{_chdir} ) {
		undef $_[0]->{_chdir};
	}
}

1;
